### PR TITLE
[FLINK-18261][parquet][orc] flink-orc and flink-parquet have invalid NOTICE file

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

flink-orc and flink-parquet provides a jar-with-dependencies.jar variant which ships binaries.
However, these binaries are not documented in META-INF/NOTICE.
There are two similar files in that directory (NOTICE from force-shading and NOTICE.txt from Commons Lang).

There is a NOTICE file that looks valid, but it is in META-INF/services.

This has been introduced in FLINK-17460.

And see FLINK-11026:

Instead of defining separate shade-plugin execution with a custom artifactSuffix this PR adds a dedicated flink-sql-** module which only contains the packaging logic. This is a similar approach that we've already been using for flink-sql-connector-elasticsearch6 .

The main motivation for this is licensing; for accurate notice files it is necessary to be able to supply each artifact with distinct NOTICE files.

This cannot be done within a single module in a reasonable way. We would have to un-package each created jar, add the appropriate license files, and re-pack them again. We'd end up with tightly-coupled plugin definitions (since the names have to match!) and an overall more complicated (and slower!) build.

## Brief change log

- Introduce flink-sql-orc to create orc bundled jar.
- Introduce flink-sql-parquet to create parquet bundled jar.
- add sqlJars profile to flink-formats to support skipping the creation of sql jars

## Verifying this change

- Manually check NOTICE file.
- Manually test bundled jar for sql filesystem connector.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-formats/flink-sql-orc/src/main/resources/META-INF/services/NOTICE [hunk 1]
```diff
@@ -1,4 +1,4 @@
-flink-orc
```
Comment (rmetzger) [inline flink-formats/flink-sql-orc/src/main/resources/META-INF/services/NOTICE:1]: The location of this file is wrong. It needs to be in the META-INF folder, not the services folder.

File: flink-formats/flink-sql-parquet/src/main/resources/META-INF/services/NOTICE [hunk 2]
```diff
@@ -1,4 +1,4 @@
-flink-parquet
```
Comment (rmetzger) [inline flink-formats/flink-sql-parquet/src/main/resources/META-INF/services/NOTICE:1]: same issue

File: flink-formats/flink-sql-orc/pom.xml [hunk 3]
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rmetzger) [inline flink-formats/flink-sql-orc/pom.xml:1]: I'm not sure if it is necessary to create new modules just to create the fat jar.
Afaik, you can use the maven shade plugin to create a new jar. I guess the `<finalName>` or `<outputFile>` parameter will be of help for this.

Comment (JingsongLi) [reply flink-formats/flink-sql-orc/pom.xml:1]: If using `<finalName>` or `<outputFile>`, I can not find a way to install or deploy into repository. >_<

Comment (rmetzger) [reply flink-formats/flink-sql-orc/pom.xml:1]: ah, so if you run "mvn install", the jar won't be added to your local `.m2` directory?

Have you tried these? http://maven.apache.org/plugins/maven-install-plugin/install-file-mojo.html (and maybe also https://maven.apache.org/plugins/maven-deploy-plugin/deploy-file-mojo.html ?)

Comment (JingsongLi) [reply flink-formats/flink-sql-orc/pom.xml:1]: My idea is to automatically upload to Maven repositories, through which users can download by themselves. (The previous jar-with-dependency didn't work too)
Like jars in https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-kafka_2.11/1.10.1/

Comment (JingsongLi) [reply flink-formats/flink-sql-orc/pom.xml:1]: I found a JIRA related to sql-jars: https://issues.apache.org/jira/browse/FLINK-11026

Comment (JingsongLi) [reply flink-formats/flink-sql-orc/pom.xml:1]: PR descriptor updated.
Profile with maven-shade-plugin and `shadedClassifierName` can create a bundled jar and upload jar to maven repositories.
But for accurate notice files it is necessary to be able to supply each artifact with distinct NOTICE files, we need create a separate module.

Comment (rmetzger) [reply flink-formats/flink-sql-orc/pom.xml:1]: > But for accurate notice files it is necessary to be able to supply each artifact with distinct NOTICE files, we need create a separate module.

But we would only need one NOTICE file in `flink-orc` which we will use for the shaded fat jar. For the regular `flink-orc.jar` we don't need a NOTICE file, as it contains only Flink binaries.

Sorry that I'm making this so complicated. I'm just trying to make sure that we are not introducing unnecessary complexity here, and that the NOTICE files are valid.

File: flink-formats/flink-sql-orc/src/main/resources/META-INF/NOTICE [hunk 4]
```diff
@@ -1,4 +1,4 @@
-flink-orc
```
Comment (rmetzger) [inline flink-formats/flink-sql-orc/src/main/resources/META-INF/NOTICE:1]: ✅  File looks good

File: flink-formats/flink-sql-parquet/src/main/resources/META-INF/NOTICE [hunk 5]
```diff
@@ -1,4 +1,4 @@
-flink-parquet
+flink-sql-parquet
```
Comment (rmetzger) [inline flink-formats/flink-sql-parquet/src/main/resources/META-INF/NOTICE:1]: ✅  file looks good

File: flink-formats/flink-sql-orc/pom.xml (hunk 3 above)
Comment (JingsongLi) [reply flink-formats/flink-sql-orc/pom.xml:1]: > Sorry that I'm making this so complicated.

It's all right, your review makes things clearer~

I thought the meaning of https://github.com/apache/flink/pull/7251 is:
For the regular `flink-orc.jar` should has a `NOTICE` file too.

If we create a `flink-sql-orc` in `flink-orc` module with classifier `sql-jar`. Then, there are two jars in artifact `flink-orc`.
And they should have different `NOTICE` files (Which is hard to do). One has many dependencies, but another just is an simple file like:
```
// ------------------------------------------------------------------
// NOTICE file corresponding to the section 4d of The Apache License,
// Version 2.0, in this case for Apache Flink
// ------------------------------------------------------------------

Apache Flink
Copyright 2006-2020 The Apache Software Foundation

This product includes software developed at
The Apache Software Foundation (http://www.apache.org/).

flink-orc
Copyright 2014-2020 The Apache Software Foundation
```

Correct me if I wrong.
CC: @zentol

Comment (rmetzger) [reply flink-formats/flink-sql-orc/pom.xml:1]: Thanks a lot for the explanation and reference. It seems that adding a second module is the way to go!

File: pom.xml [hunk 6]
```diff
@@ -156,6 +156,7 @@ under the License.
 		<hivemetastore.hadoop.version>2.7.5</hivemetastore.hadoop.version>
 		<japicmp.referenceVersion>1.10.0</japicmp.referenceVersion>
 		<japicmp.outputDir>tools/japicmp-output</japicmp.outputDir>
+		<flink.format.parquet.version>1.10.0</flink.format.parquet.version>
```
Comment (zentol) [inline pom.xml:159]: I'd prefer this to be defined in `flink-formats`; to keep things as localized as possible.

**Comment by JingsongLi** [comment]
Thanks @rmetzger and @zentol for review, merging...

**Pull request merged**